server/php/index.php

```diff
@@ -0,0 +1,198 @@
+    'country' => $config['accountCountry'],
+    'currency' => $config['accountCurrency'],
+    'paymentMethods' => implode($config['paymentMethods'], ', '),
+    'shippingOptions' => Shipping::getShippingOptions()
+  ]);
+});
+
+// List of fake products on our fake shop
+// Used to display the user's cart and calculate the total price
+$app->get('/products', function (Request $request, Response $response, array $args) {
+  return $response->withJson(Inventory::listProducts());
+});
+
+// List of fake products on our fake shop
+// Used to display the user's cart and calculate the total price
+$app->get('/products/{id}', function (Request $request, Response $response, array $args) {
+  return $response->withJson(Inventory::getProduct($args['id']));
+});
+
+// Create the payment intent
+// Used when the user starts the checkout flow
+$app->post('/payment_intents', function (Request $request, Response $response, array $args) {
+  $data = $request->getParsedBody();
+  try {
+    $paymentIntent = \Stripe\PaymentIntent::create([
+      'amount' => Inventory::calculatePaymentAmount($data['items']),
+      'currency' => $data['currency'],
+      'payment_method_types' => $this->get('settings')['stripe']['paymentMethods']
+    ]);
+
+    return $response->withJson([ 'paymentIntent' => $paymentIntent ]);
+  } catch (\Exception $e) {
+    return $response->withJson([ 'error' => $e->getMessage() ])->withStatus(403);
+  }
+});
+
+// Update the total when selected a different shipping option via the payment request API
+$app->post('/payment_intents/{id}/shipping_change', function (Request $request, Response $response, array $args) {
+  $data = $request->getParsedBody();
+  $amount = Inventory::calculatePaymentAmount($data['items']);
+  $amount += Shipping::getShippingCost($data['shippingOption']['id']);
+  try {
+    $paymentIntent = \Stripe\PaymentIntent::update($args['id'], [ 'amount' => $amount ]);
+    return $response->withJson([ 'paymentIntent' => $paymentIntent ]);
+  } catch (\Exception $e) {
+    return $response->withJson([ 'error' => $e->getMessage() ])->withStatus(403);
+  }
+});
+
+// Fetch the payment intent status
+// Used for redirect sources when coming back to the return URL
+$app->get('/payment_intents/{id}/status', function (Request $request, Response $response, array $args) {
+    $paymentIntent = \Stripe\PaymentIntent::retrieve($args['id']);
+    return $response->withJson([ 'paymentIntent' => [ 'status' => $paymentIntent->status ] ]);
+});
+
+// Events receiver for payment intents and sources
+$app->post('/webhook', function (Request $request, Response $response, array $args) {
+  $logger = $this->get('logger');
+
+  $event = $request->getParsedBody();
+
+  // Parse the message body (and check the signature if possible)
+  $webhookSecret = $this->get('settings')['stripe']['webhookSecret'];
+  if ($webhookSecret) {
+    try {
+      $event = \Stripe\Webhook::constructEvent(
+        @file_get_contents('php://input'),
```

Comment:
I know we have to get the raw request body here(and not a parsed version that Slim may be putting into the the `$request` variable, like `getParsedBody`), or the signatures won't match. Is there maybe a more idiomatic way to do that in Slim? Would `$request->getContent()` or [similar](https://www.slimframework.com/docs/v3/objects/response.html#the-response-body) work? Feel free to ignore this, I'm just wondering if we can also use this as a nice example of how to get a raw body for a webhook in this framework.

Reply:
Indeed, replaced with $request->getBody()